src/Algebra/Module/Construct/Nagata.agda

```diff
@@ -0,0 +1,209 @@
+-- "Forward- or reverse-mode automatic differentiation: What's the difference?"
+-- (Van den Berg, Schrijvers, McKinna, Vandenbroucke;
+-- Science of Computer Programming, Vol. 234, January 2024
+-- https://doi.org/10.1016/j.scico.2023.103010)
+--
+-- The construction N =def R ⋉ M , for which there is unfortunately
+-- no consistent notation in the literature, consists of:
+-- * carrier: pairs |R| × |M|
+-- * with additive structure that of the direct sum R ⊕ M _of modules_
+-- * but with multiplication _*_ such that M forms an _ideal_ of N
+-- * moreover satisfying 'm * m ≈ 0' for every m ∈ M ⊆ N
+--
+-- The fundamental lemma (proved here) is that N, in fact, defines a Ring:
+-- this ring is essentially the 'ring of dual numbers' construction R[M]
+-- (Clifford, 1874; generalised!) for an ideal M, and thus the synthetic/algebraic
+-- analogue of the tangent space of M (considered as a 'vector space' over R)
+-- in differential geometry, hence its application to Automatic Differentiation.
+--
+-- Nagata's more fundamental insight (not yet shown here) is that
+-- the lattice of R-submodules of M is in order-isomorphism with
+-- the lattice of _ideals_ of R ⋉ M , and hence that the study of
+-- modules can be reduced to that of ideals of a ring, and vice versa.
+--
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Algebra.Module.Construct.Nagata where
+
+open import Algebra.Bundles using (AbelianGroup; Ring)
+open import Algebra.Core
+import Algebra.Consequences.Setoid as Consequences
+import Algebra.Definitions as Definitions
+open import Algebra.Module.Bundles using (Bimodule)
+import Algebra.Module.Construct.DirectProduct as DirectProduct
+import Algebra.Module.Construct.TensorUnit as TensorUnit
+open import Algebra.Structures using (IsAbelianGroup; IsRing)
+open import Data.Product using (_,_; ∃-syntax)
+open import Level using (Level; _⊔_)
+open import Relation.Binary using (Rel; Setoid; IsEquivalence)
+import Relation.Binary.Reasoning.Setoid as ≈-Reasoning
+
+private
+  variable
+    m ℓm r ℓr : Level
+
+------------------------------------------------------------------------
+-- Definitions
+
+module Nagata (ring : Ring r ℓr) (bimodule : Bimodule ring ring m ℓm) where
+
+  private
+    open module R = Ring ring
+      using ()
+      renaming (Carrier to R)
+
+    open module M = Bimodule bimodule
+      renaming (Carrierᴹ to M)
+
+    open AbelianGroup M.+ᴹ-abelianGroup
+      renaming (setoid to setoidᴹ; sym to symᴹ)
+
+    +ᴹ-middleFour = Consequences.comm∧assoc⇒middleFour setoidᴹ +ᴹ-cong +ᴹ-comm +ᴹ-assoc
+
+    open ≈-Reasoning setoidᴹ
+
+    open module N = Bimodule (DirectProduct.bimodule TensorUnit.bimodule bimodule)
+      using ()
+      renaming (Carrierᴹ to N
+               ; _≈ᴹ_ to _≈ᴺ_
+               ; _+ᴹ_ to _+_
+               ; 0ᴹ to 0#
+               ; -ᴹ_ to -_
+               ; +ᴹ-isAbelianGroup to +-isAbelianGroup
+               )
```

Comment:
So I'm still doing this bit of renaming, so that later the types of the various `(Is)Ring` operations/properties are standard for that signature (rather than a lot of duplication of eg. `N._+ᴹ_` in place of `_+_`... etc.)

Reply:
This is a much more 'controlled' but of renaming that seems to make sense.